Version control data source

lukedirtwalker opened the pull request:
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23discussion_r41978439%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23discussion_r41978510%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23discussion_r41978646%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23issuecomment-148010761%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23issuecomment-148011161%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23discussion_r41979721%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23discussion_r41980582%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23issuecomment-148023298%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23issuecomment-148010761%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Nice%2C%20only%20two%20minor%20issues.%20Look%20at%20those%20and%20I%27ll%20merge%20it.%22%2C%20%22created_at%22%3A%20%222015-10-14T10%3A45%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%27s%20also%20nice%20to%20see%20that%20this%20was%20rather%20straightforward%20to%20do.%20Among%20other%20things%2C%20this%20means%20that%20the%20queries%20system%20is%20well%20designed%20and%20easy%20to%20extend.%20%3A%29%22%2C%20%22created_at%22%3A%20%222015-10-14T10%3A46%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed%20the%202%20small%20issues.%20User%20error%20feedback%20is%20definitely%20for%20later%20%28we%20already%20discussed%20a%20solution%20btw%29.%22%2C%20%22created_at%22%3A%20%222015-10-14T11%3A30%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%200b28104c39021620bb1615f62fb32672fdc6392c%20InformationScripting/src/queries/VersionControlQuery.cpp%2058%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23discussion_r41978510%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20not%20simply%3A%20%60GitRepository%20repository%7Bpath%7D%60.%20Why%20do%20we%20need%20a%20unique%20pointer%3F%22%2C%20%22created_at%22%3A%20%222015-10-14T10%3A42%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/VersionControlQuery.cpp%3AL1-116%22%7D%2C%20%22Pull%200b28104c39021620bb1615f62fb32672fdc6392c%20InformationScripting/src/queries/VersionControlQuery.cpp%2056%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23discussion_r41978439%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Hmm%2C%20at%20some%20point%20we%20should%20actually%20come%20up%20with%20a%20mechanism%20to%20report%20these%20errors.%22%2C%20%22created_at%22%3A%20%222015-10-14T10%3A41%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/VersionControlQuery.cpp%3AL1-116%22%7D%2C%20%22Pull%200b28104c39021620bb1615f62fb32672fdc6392c%20InformationScripting/src/queries/VersionControlQuery.cpp%2073%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/168%23discussion_r41978646%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Or%20modified.%20Renaming%20a%20class%20for%20example%20is%20a%20stationary%20change%2C%20but%20there%20are%20no%20added/deleted%20children.%22%2C%20%22created_at%22%3A%20%222015-10-14T10%3A44%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22The%20problem%20here%20is%20a%20child%20change%20is%20detected%20as%20stationary%20change%20as%20well.%20%5Cr%5CnFor%20example%20if%20I%20change%20a%20statement%20in%20a%20StatementItemList%20the%20list%20will%20appear%20as%20a%20stationary%20change%2C%20which%20I%20don%27t%20like%20for%20this%20purpose.%22%2C%20%22created_at%22%3A%20%222015-10-14T10%3A58%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22But%20there%20is%20a%20way%20to%20check%20for%20that%20specifically.%20use%20%60onlyStructureChange%60%20and%20discard%20changes%20where%20this%20is%20true.%22%2C%20%22created_at%22%3A%20%222015-10-14T11%3A09%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/VersionControlQuery.cpp%3AL1-116%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-Pull 0b28104c39021620bb1615f62fb32672fdc6392c InformationScripting/src/queries/VersionControlQuery.cpp 56'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/168#discussion_r41978439'>File: InformationScripting/src/queries/VersionControlQuery.cpp:L1-116</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Hmm, at some point we should actually come up with a mechanism to report these errors.
- [x] <a href='#crh-comment-Pull 0b28104c39021620bb1615f62fb32672fdc6392c InformationScripting/src/queries/VersionControlQuery.cpp 58'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/168#discussion_r41978510'>File: InformationScripting/src/queries/VersionControlQuery.cpp:L1-116</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Why not simply: `GitRepository repository{path}`. Why do we need a unique pointer?
- [x] <a href='#crh-comment-Pull 0b28104c39021620bb1615f62fb32672fdc6392c InformationScripting/src/queries/VersionControlQuery.cpp 73'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/168#discussion_r41978646'>File: InformationScripting/src/queries/VersionControlQuery.cpp:L1-116</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Or modified. Renaming a class for example is a stationary change, but there are no added/deleted children.
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> The problem here is a child change is detected as stationary change as well.
  For example if I change a statement in a StatementItemList the list will appear as a stationary change, which I don't like for this purpose.
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> But there is a way to check for that specifically. use `onlyStructureChange` and discard changes where this is true.
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/168#issuecomment-148010761'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Nice, only two minor issues. Look at those and I'll merge it.
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> It's also nice to see that this was rather straightforward to do. Among other things, this means that the queries system is well designed and easy to extend. :)
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Fixed the 2 small issues. User error feedback is definitely for later (we already discussed a solution btw).

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/168?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/168?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/168'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

lukedirtwalker committed:
Build lookup hash before calling find on GenericTree

lukedirtwalker committed:
Fix a memory leak in GitRepository

lukedirtwalker committed:
Fix wrong initialization in HighlightOverlay

lukedirtwalker committed:
Prevent division by zero in Heatmap

lukedirtwalker committed:
Initial implemenation of a version control data source / query

lukedirtwalker committed:
Add count argument to VersionControlQuery

dimitar-asenov commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/168'>Refresh</a>.<a href='#crh-update-none'></a>

dimitar-asenov commented on InformationScripting/src/queries/VersionControlQuery.cpp:
```diff
@@ -0,0 +1,115 @@
+	// get GitRepository
+	QString path("projects/");
+	path.append(managerName);
+	// TODO user error later:
```
Hmm, at some point we should actually come up with a mechanism to report these errors.

dimitar-asenov commented on InformationScripting/src/queries/VersionControlQuery.cpp:
```diff
@@ -0,0 +1,115 @@
+	path.append(managerName);
+	// TODO user error later:
+	Q_ASSERT(GitRepository::repositoryExists(path));
+	auto repository = std::make_unique<GitRepository>(path);
```
Why not simply: `GitRepository repository{path}`. Why do we need a unique pointer?

dimitar-asenov commented on InformationScripting/src/queries/VersionControlQuery.cpp:
```diff
@@ -0,0 +1,115 @@
+
+		for (auto change : changes.values())
+		{
+			// We ignore stationary changes since a node with a stationary change, has a child which was deleted or added,
```
Or modified. Renaming a class for example is a stationary change, but there are no added/deleted children.

lukedirtwalker replied:
The problem here is a child change is detected as stationary change as well. 
For example if I change a statement in a StatementItemList the list will appear as a stationary change, which I don't like for this purpose.

dimitar-asenov replied:
But there is a way to check for that specifically. use `onlyStructureChange` and discard changes where this is true.

dimitar-asenov commented:
Nice, only two minor issues. Look at those and I'll merge it.

dimitar-asenov commented:
It's also nice to see that this was rather straightforward to do. Among other things, this means that the queries system is well designed and easy to extend. :)

lukedirtwalker committed:
Dont't ignore Stationary changes but structural changes in VC-Query

This was actually the intention in the first place, but I din't find the
correct function.

lukedirtwalker committed:
Use stack allocated GitRepository instead of unique_ptr

There is no reason to allocate it dynamically.

lukedirtwalker commented:
Fixed the 2 small issues. User error feedback is definitely for later (we already discussed a solution btw).
